enterprise_gateway/services/kernels/handlers.py

```diff
@@ -146,11 +153,178 @@ def get(self, kernel_id):
+            raise web.HTTPError(400, "Missing required field `env` in payload.")
+        kernel = km.get_kernel(kernel_id)
+        if kernel.restarting:  # handle duplicate request.
+            self.log.info(
+                "An existing restart request is still in progress. Skipping this request."
+            )
+            raise web.HTTPError(
+                400, "Duplicate Kernel update request received for Id: %s." % kernel_id
+            )
+        try:
+            # update Kernel metadata
+            kernel.set_user_extra_overrides(payload)
+            await km.restart_kernel(kernel_id)
+            kernel.fire_kernel_event_callbacks(
+                event="kernel_refresh", zmq_messages=payload.get("zmq_messages", {})
+            )
+        except web.HTTPError as he:
+            self.log.exception(
+                f"HTTPError exception occurred in refreshing kernel: {kernel_id}: {he}"
+            )
+            await km.shutdown_kernel(kernel_id)
+            kernel.fire_kernel_event_callbacks(
+                event="kernel_refresh_failure", zmq_messages=payload.get("zmq_messages", {})
+            )
+            raise he
+        except Exception as e:
+            self.log.exception(f"An exception occurred in updating kernel : {kernel_id}: {e}")
+            await km.shutdown_kernel(kernel_id)
+            kernel.fire_kernel_event_callbacks(
+                event="kernel_refresh_failure", zmq_messages=payload.get("zmq_messages", {})
+            )
+            raise web.HTTPError(
+                500,
+                "Error occurred while refreshing Kernel: %s." % kernel_id,
+                reason=f"{e}",
+            )
+        else:
+            response_body = {"message": f"Successfully refreshed kernel with id: {kernel_id}"}
+            self.finish(json.dumps(response_body, default=date_default))
+            return
+
+
+class RemoteZMQChannelsHandler(
+    TokenAuthorizationMixin, CORSMixin, JSONErrorsMixin, jupyter_server_handlers.ZMQChannelsHandler
+):
+    def open(self, kernel_id):
+        self.log.info(f"Websocket open request received for kernel: {kernel_id}")
+        super().open(kernel_id)
+        km = self.kernel_manager
+        km.add_kernel_event_callbacks(kernel_id, self.on_kernel_refresh, "kernel_refresh")
+        km.add_kernel_event_callbacks(
+            kernel_id, self.on_kernel_refresh_failure, "kernel_refresh_failure"
+        )
+
+    def on_kernel_refresh(self, **kwargs):
+        self.log.info("Refreshing the client websocket to kernel connection.")
+        self.refresh_zmq_sockets()
+        zmq_messages = kwargs.get("zmq_messages", {})
+        if "stream_reply" in zmq_messages:
+            self.log.debug("Sending stream_reply success message.")
+            success_message = zmq_messages.get("stream_reply")
+            success_message["content"] = {
+                "name": "stdout",
+                "text": "The Kernel is successfully refreshed.",
```

Comment:
This is throughout.  Please lowercase `kernel` unless part of a name (e.g., `KernelManager`).